<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PhageHosts &mdash; PhageHosts 1.0 documentation</title>
    
    <link rel="stylesheet" href="static/better.css" type="text/css" />
    <link rel="stylesheet" href="static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="static/jquery.js"></script>
    <script type="text/javascript" src="static/underscore.js"></script>
    <script type="text/javascript" src="static/doctools.js"></script>
    <link rel="top" title="PhageHosts 1.0 documentation" href="#" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head>
  <body role="document">
    <header id="pageheader"><h1><a href="# ">
        PhageHosts 1.0 documentation
    </a></h1></header>

<div style="background-color: white; text-align: left; padding: 10px 10px 15px 15px">
	<a href="#"><img src="static/phagehost.png" border="0" alt="sampledoc"/></a>
</div>

  <div class="related top">
  <nav id="rellinks">
    <ul>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="#">Phage Host Analysis</a></li> 
    </ul>
  </nav>
  </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">PhageHosts</a></li>
<li><a class="reference internal" href="#datasets">Datasets</a><ul>
<li><a class="reference internal" href="#phages">Phages</a><ul>
<li><a class="reference internal" href="#extracting-the-dna-and-protein-sequences">Extracting the DNA and protein sequences</a></li>
</ul>
</li>
<li><a class="reference internal" href="#bacteria">Bacteria</a></li>
<li><a class="reference internal" href="#extracting-taxonomy-information">Extracting taxonomy information</a><ul>
<li><a class="reference internal" href="#resulting-files">Resulting files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#scoring-predictions">Scoring predictions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#genetic-homology">1. Genetic Homology</a><ul>
<li><a class="reference internal" href="#comparing-the-phages-against-the-complete-bacterial-genomes-blastx">Comparing the phages against the complete bacterial genomes (blastx)</a></li>
<li><a class="reference internal" href="#comparing-phages-to-the-non-redundant-nr-database">Comparing phages to the non-redundant (nr) database.</a></li>
<li><a class="reference internal" href="#similarity-to-complete-genomes-blastn">Similarity to complete genomes (blastn)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#crispr-spacers-crispr-spacers">2. CRISPR spacers. CRISPR spacers</a></li>
<li><a class="reference internal" href="#exact-matches">3. Exact matches</a></li>
<li><a class="reference internal" href="#oligonucleotide-profiles">4. Oligonucleotide profiles</a><ul>
<li><a class="reference internal" href="#gc-content-of-coding-regions">GC Content of Coding Regions</a></li>
<li><a class="reference internal" href="#codon-usage">Codon usage</a></li>
<li><a class="reference internal" href="#kmer-profiles">Kmer profiles</a></li>
</ul>
</li>
<li><a class="reference internal" href="#cooccurrence-analysis">5. Cooccurrence analysis</a><ul>
<li><a class="reference internal" href="#predicting-bacterial-presence">Predicting bacterial presence</a></li>
<li><a class="reference internal" href="#predicting-phage-sequences">Predicting phage sequences</a></li>
<li><a class="reference internal" href="#mutual-information">Mutual Information</a></li>
</ul>
</li>
<li><a class="reference internal" href="#you-can-also-find-what-you-are-looking-for-directly">You can also find what you are looking for directly:</a></li>
</ul>

<div class="sourcelink">
  <a href="sources/index.txt" rel="nofollow">
    Show page source
  </a>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>

        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="phagehosts">
<h1>PhageHosts<a class="headerlink" href="#phagehosts" title="Permalink to this headline">¶</a></h1>
<p>The code used for identification of the hosts of different phages. This
is the complete code base used in *Robert A. Edwards, Katelyn McNair,
Karoline Faust, Jeroen Raes, and Bas E. Dutilh (2015) Computational
approaches to predict bacteriophage–host relationships. FEMS
Microbiology Reviews <a class="reference external" href="http://dx.doi.org/10.1093/femsre/fuv048">doi:
10.1093/femsre/fuv048</a>∗</p>
<p>Almost all of the code is written in python and should work with version
2.7. Some parts of the code require matplotlib, numpy, and/or scipy.
Some parts of the code were written in Perl 5 and should run with the
standard Perl libraries. The code was written on CentOS 6 machines and
should run <em>out of the box</em> on those machines, but it will also run on
Linux, MacOSX, or Windows. Parts of the code include reference to
running things on our cluster where we use SGE as the job scheduler. You
can run those parts without the cluster, but it will probably be slower!</p>
<p>The data/ directory includes some of the data sets that we used. We have
not uploaded all genome sequences: you can get those from
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a>.</p>
</div>
<div class="section" id="datasets">
<h1>Datasets<a class="headerlink" href="#datasets" title="Permalink to this headline">¶</a></h1>
<div class="section" id="phages">
<h2>Phages<a class="headerlink" href="#phages" title="Permalink to this headline">¶</a></h2>
<p>The phage genomes were downloaded from RefSeq and parsed to get the
information in the &#8220;host&#8221; field. These data were converted to coding and
protein sequences:</p>
<p>For example, to convert from GBFF to FNA format for all of the open
reading frames:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/combine_gbff_fna.py viral.1.genomic.gbff viral.1.1.genomic.fna &gt; viral.1.cds.fna
</pre></div>
</div>
<p>There are 971 phages in RefSeq that have a host annotation:</p>
<div class="highlight-python"><div class="highlight"><pre>perl -ne &#39;@a=split /\t/; print if ($a[2])&#39; refseq.txt | grep YES$ | grep -i &#39;complete genome&#39; | wc -l
</pre></div>
</div>
<p>use <code class="docutils literal"><span class="pre">perl</span> <span class="pre">PhageHosts/code/get_viral_dna.pl</span></code> to split the viruses into
either phage or eukaryotic viruses. For this work we are just going to
use the Phage datasets. It is left up to the reader to try some of these
challenges on Eukaryotic viral data sets.</p>
<div class="section" id="extracting-the-dna-and-protein-sequences">
<h3>Extracting the DNA and protein sequences<a class="headerlink" href="#extracting-the-dna-and-protein-sequences" title="Permalink to this headline">¶</a></h3>
<p>To get the phage coding sequences, we pull them out of the fasta file,
and then translate them</p>
<div class="highlight-python"><div class="highlight"><pre>for i in $(cut -f 1 phage_with_host.tsv); do
    grep -A 1 $i 2014_07_21/refseq/viral.1.cds.fna; done &gt; data/phage_with_host.cds.fna
    perl PhageHosts/code/translate.pl phage_with_host.cds.fna &gt; data/phage_with_host.cds.faa
done
</pre></div>
</div>
</div>
</div>
<div class="section" id="bacteria">
<h2>Bacteria<a class="headerlink" href="#bacteria" title="Permalink to this headline">¶</a></h2>
<p>All bacterial sequences were downloaded from RefSeq:</p>
<div class="highlight-python"><div class="highlight"><pre>ncftpget ftp://ftp.ncbi.nih.gov/refseq/release/bacteria/bacteria.*.fna.gz
</pre></div>
</div>
<p>These were extracted and a single bacterial database was made for blastn
searches</p>
<div class="highlight-python"><div class="highlight"><pre>cat bacteria.*fna &gt; bacteria.genomic.fna
</pre></div>
</div>
<p>We extract the NC_ ids from the complete bacteria, above:</p>
<div class="highlight-python"><div class="highlight"><pre>grep \&gt; bacteria.genomic.fna &gt; ids.txt
perl -i -npe &#39;s/\:/\t/;s/^(.*)\|\s+/$1\|\t/&#39; ids.txt
</pre></div>
</div>
<p>Then trim these down to complete genomes:</p>
<div class="highlight-python"><div class="highlight"><pre>egrep &#39;complete genome|complete sequence&#39; ids.txt | grep -v plasmid | grep -v &#39;whole genome shotgun sequence&#39; | grep -v NR_ &gt; complete_genome_ids.txt
</pre></div>
</div>
<p>We edited this last file manually to ensure that we only have complete
bacterial genomes.</p>
<p>We also downloaded the gbff files and orf files from RefSeq, and then
used those to create a tbl file with both protein and CDS sequences, and
create file called protein_list that has a list of all of the gbff
files that we extracted.</p>
<div class="highlight-python"><div class="highlight"><pre>F=$(head -n $SGE_TASK_ID protein_list | tail -n 1)
UF=$(echo $F | sed -e &#39;s/.gz//&#39;)
gunzip $F
perl PhageHosts/code/genbank2flatfile.pl $UF
gzip $UF
</pre></div>
</div>
<p>Based on this output we create two files, refseq_proteins.faa (with
proteins) and refseq_orfs.faa (with DNA) from just the complete
genomes.</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/tbl2protdna.py .
</pre></div>
</div>
</div>
<div class="section" id="extracting-taxonomy-information">
<h2>Extracting taxonomy information<a class="headerlink" href="#extracting-taxonomy-information" title="Permalink to this headline">¶</a></h2>
<p>We wrote code to automatically get the taxonomy for most of the hosts
from the RefSeq files, but there were a few that we could not map, so we
added those manually. To whit:</p>
<div class="highlight-python"><div class="highlight"><pre>&#39;Acinetobacter genomosp.&#39; : &#39;471&#39;,
&#39;Actinobacillus actinomycetemcomitans&#39; : &#39;714&#39;,
&#39;alpha proteobacterium&#39; : &#39;34025&#39;,
&#39;Bacillus clarkii&#39; : &#39;79879&#39;,
&#39;Brevibacterium flavum&#39; : &#39;92706&#39;,
&#39;Celeribacter sp.&#39; : &#39;875171&#39;,
&#39;Escherichia sp.&#39; : &#39;237777&#39;,
&#39;Geobacillus sp.&#39; : &#39;340407&#39;,
&#39;Gordonia rubropertincta&#39; : &#39;36822&#39;,
&#39;Iodobacter sp.&#39; : &#39;641420&#39;,
&#39;Listeria sp.&#39; : &#39;592375&#39;,
&#39;Marinomonas sp.&#39; : &#39;127794&#39;,
&#39;Methanobacterium thermoautotrophicum&#39; : &#39;145262&#39;,
&#39;methicillin-resistant Staphylococcus&#39; : &#39;1280&#39;,
&#39;Nitrincola sp.&#39; : &#39;459834&#39;,
&#39;Persicivirga sp.&#39; : &#39;859306&#39;,
&#39;Salisaeta sp.&#39; : &#39;1392396&#39;,
&#39;Sulfitobacter sp.&#39; : &#39;191468&#39;
</pre></div>
</div>
<p>Then we generate the tsv file:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/phage_host_taxonomy.py  &gt; phage_host_taxonomy.tsv.
</pre></div>
</div>
<p>We also just made two files with tuples of genome NC id and taxonomy id.</p>
<p><strong>NOTE:</strong> <em>When we score the connection between phage and host, we need
to know the taxonomy ID of the host, not that of the phage, to see if
there is a match. Therefore, in this file the taxonomy ID is that of the
**host*</em> (not the phage!)*</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/phage2taxonomy.py  &gt; phage_host_taxid.txt
</pre></div>
</div>
<p>and another file refseq2taxonomy.py which added the tax id to the list
of complete bacterial genomes, and then we made a list of bacteria and
taxid:</p>
<div class="highlight-python"><div class="highlight"><pre>cut -f 2,4 /lustre/usr/data/NCBI/RefSeq/bacteria/complete_genome_ids_taxid.txt | perl -pe &#39;s/^.*\|N/N/; s/\.\d+\|//&#39; &gt; bacteria_taxid.txt
</pre></div>
</div>
<p>We trimmed out any phages that we can not match at the species level:</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 PhageHosts/code/comparePhageToHosts.py
</pre></div>
</div>
<p>Then we combined those into a single file for all tax ids</p>
<div class="highlight-python"><div class="highlight"><pre>cat phage_taxid.txt bacteria_taxid.txt &gt; data/all_host_taxid.txt
</pre></div>
</div>
<p>And add the taxonomy to those files:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/phage_host_add_taxonomy.py all_host_taxid.txt &gt; data/all_host_taxid_taxonomy.txt
</pre></div>
</div>
<p>Note that in this process we deleted two phages whose hosts were not
really known (NC_000935) APSE-1 whose host was Endosymbiont and Zamilon
virophage (NC_022990) whose host is Mont1 megavirus)</p>
<div class="section" id="resulting-files">
<h3>Resulting files<a class="headerlink" href="#resulting-files" title="Permalink to this headline">¶</a></h3>
<p>This resulted in one key file that we use in this work
<a class="reference external" href="https://github.com/linsalrob/PhageHosts/blob/master/data/all_host_taxid.txt">data/all_host_taxid.txt</a>
which just has two columns, a RefSeq ID (which we sometimes call NC id)
and a taxonomy ID. When the RefSeq ID refers to a bacterial sequence,
the taxonomy ID is of the bacterium from where the sequence came. When
the RefSeq ID refers to a phage sequence, the taxonomy ID is of the
phage&#8217;s host.</p>
</div>
</div>
<div class="section" id="scoring-predictions">
<h2>Scoring predictions<a class="headerlink" href="#scoring-predictions" title="Permalink to this headline">¶</a></h2>
<p>In general, we are going to print out tab separated files containing the
phage ID in the first column, and the ID of any potential hosts in
subsequent columns. We do not necessarily know how many columns there
will be. This is a flexible format that allows us to convert those IDs
to taxonomy IDs, and then use the NCBI hierarchy to move through the
phylogenetic tree to score how good our matches are.</p>
<p>We have a few keys pieces of code for this work:</p>
<ul class="simple">
<li>NC2taxid.py is python code that converts any RefSeq ID (NC_\d+)
into its associated taxonomy ID based on the association above, the
bacterial RefSeq IDs map to bacterial taxonomy IDs and the phage
RefSeq IDs map to host taxonomy IDs.</li>
<li>scoreTaxId.py is python code that takes the a set of taxonomy IDs and
scores all subsequent elements in the set against the first member of
the set, using &#8216;species&#8217;, &#8216;genus&#8217;, &#8216;family&#8217;, &#8216;order&#8217;, &#8216;class&#8217;,
&#8216;phylum&#8217;, &#8216;superkingdom&#8217; taxonomic levels from NCBI.</li>
</ul>
</div>
</div>
<div class="section" id="genetic-homology">
<h1>1. Genetic Homology<a class="headerlink" href="#genetic-homology" title="Permalink to this headline">¶</a></h1>
<div class="section" id="comparing-the-phages-against-the-complete-bacterial-genomes-blastx">
<h2>Comparing the phages against the complete bacterial genomes (blastx)<a class="headerlink" href="#comparing-the-phages-against-the-complete-bacterial-genomes-blastx" title="Permalink to this headline">¶</a></h2>
<p>Start by making a database of just the complete genomes protein
sequences</p>
<div class="highlight-python"><div class="highlight"><pre>PhageHosts/code/refseq2complete.py $HOME/phage/host_analysis/bacteria_taxid.txt  refseq_proteins.faa complete_genome_proteins.faa
</pre></div>
</div>
<p>and then blast those:</p>
<div class="highlight-python"><div class="highlight"><pre>PhageHosts/code/split_blast_queries_edwards_blastplus.pl -f ../../phage_with_host.fna -n 100 -p blastx -d phage.blastx -db RefSeq/bacteria/complete_genome_proteins.faa -evalue 0.01
cat phage.blastx/*blastx &gt; phage.complete.blastx
</pre></div>
</div>
<p><em>NOTE:</em> There are mutliple proteins with the same ID that come from
different genomes, however these are identical proteins (at the amino
acid level). Therefore the blastx searches all return the same results
for each protein. I use this one line of Perl code to print out unique
solutions from the blastx output.</p>
<div class="highlight-python"><div class="highlight"><pre>perl -ne &#39;print unless ($s{$_}); $s{$_}=1&#39; phage.complete.blastx &gt; phage.complete.unique.blastx
</pre></div>
</div>
<p>We convert the output so we just have NC identifiers:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/blastx2genome.py phage.complete.unique.blastx phage.genomes.blastx
</pre></div>
</div>
<p>Now we just count the hosts with the most number of hits to each of the
phages, and score those hits</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/mostBlastHits.py phage.genomes.blastx &gt; most.tax
python2.7 PhageHosts/code/scoreTaxId.py most.tax &gt; score.tax
</pre></div>
</div>
</div>
<div class="section" id="comparing-phages-to-the-non-redundant-nr-database">
<h2>Comparing phages to the non-redundant (nr) database.<a class="headerlink" href="#comparing-phages-to-the-non-redundant-nr-database" title="Permalink to this headline">¶</a></h2>
<p>To see what would happen, we also compared the phages to all the
proteins in the non-redundant database. Without going into the results
in any detail, they weren&#8217;t as good as using the complete genomes
because there are more proteins in the nr database and thus more
confusion from the similarity searches.</p>
<p>We used blast to compare the complete phage genomes against all
bacterial proteins in the GenBank nr and then gi_taxid table of the
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/">NCBI Taxonomy Site</a> (<a class="reference external" href="http://tinyurl.com/ncbiftp">here
is an alternate link</a> as there is an
issue linking to FTP sites from github) to get the taxonomy id of the
top hits. We compared those hits with the expected hosts.</p>
<p>The blast is a standard blast using nr as the database, and our output
file is in blast tab separated output format (standard format with
<em>qlen</em> and <em>slen</em> added to the output). We then use</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/blast2taxid.py  phage_with_host.fna.blastx prot &gt; phage.host.blastx.taxid
</pre></div>
</div>
<p>to create a list of phage and host taxids. Now we have to convert these
to taxids and score the hits:</p>
<div class="highlight-python"><div class="highlight"><pre>for i in all equal best;
    do echo $i;
    python PhageHosts/code/blast_hits.py phage.host.blastx.taxid $i ${i}_hits.txt;
    python2.7 PhageHosts/code/scoreTaxId.py ${i}_hits.txt &gt; ${i}_hits.score;
done
</pre></div>
</div>
<p>This creates three files, <code class="docutils literal"><span class="pre">all_hits.score</span></code>, <code class="docutils literal"><span class="pre">equal_hits.score</span></code>,
<code class="docutils literal"><span class="pre">best_hits.score</span></code>.</p>
</div>
<div class="section" id="similarity-to-complete-genomes-blastn">
<h2>Similarity to complete genomes (blastn)<a class="headerlink" href="#similarity-to-complete-genomes-blastn" title="Permalink to this headline">¶</a></h2>
<p>We start with direct matches between the phages and hosts using the
complete bacterial genomic DNA, first by making a database and then by
using our cluster to blast the phages against the database. <em>Note:</em> We
change the default values here to make them the similar as NCBI blast
for a whole genomes, however that usually uses a word size of 28. I cut
the word size to 5 so we get some shorter matches. Alternate parameters
will affect the results!</p>
<div class="highlight-python"><div class="highlight"><pre>make a blast db: makeblastdb -in bacteria.genomic.fna -dbtype nucl
PhageHosts/code/split_blast_queries_edwards_blastplus.pl -f phage_with_host.fna -n 100 -db /lustre/usr/data/NCBI/RefSeq/bacteria/complete_genomes.fna -d phage.blastn -p blastn -word_size 5 -evalue 10 -num_descriptions 100 -reward 1 -penalty -2 -gapopen 0 -gapextend 0 -outfmt &#39;6 std qlen slen&#39;
cat phage.blastn/blastn &gt;&gt; phage_host.blastn
</pre></div>
</div>
<p>We then figure out the best hit for each phage, and score those hits.</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 PhageHosts/code/parse_blastn.py phage_host.blastn &gt; phage.hits
python PhageHosts/code/NC2taxid.py phage.hits &gt; phage.taxid
python2.7 PhageHosts/code/scoreTaxId.py phage.taxid &gt; blastn.score
</pre></div>
</div>
<p>To make a table of all the scores between all phage genomes and all bacterial genomes we use</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 ../GitHubRepository/PhageHosts/code/blastn_all_v_all.py phage_host.blastn
</pre></div>
</div>
</div>
</div>
<div class="section" id="crispr-spacers-crispr-spacers">
<h1>2. CRISPR spacers. CRISPR spacers<a class="headerlink" href="#crispr-spacers-crispr-spacers" title="Permalink to this headline">¶</a></h1>
<p>Pilercr was downloaded from <a class="reference external" href="http://www.drive5.com/pilercr/">drive5</a>
and run against all complete genomes to create the database
<a class="reference external" href="data/pilercr1.06.output.fna">data/pilercr1.06.output.fna</a>.</p>
<p>blastn searches were performed using modified parameters that should be
better for short matches:</p>
<div class="highlight-python"><div class="highlight"><pre>/usr/local/blast+/bin/blastn -db pilercr1.06.output.fna -query phage_with_host.fna -out phage_with_host.fna.crispr.blastn -outfmt &#39;6 std qlen slen&#39; -evalue 1 -gapopen 10 -penalty -1 -gapextend 2 -word_size 7 -dust no -task blastn-short
</pre></div>
</div>
<p>These blast searches were
<a class="reference external" href="PhageHosts/code/score_blast.py%20score_blast.py">compared</a>,
<a class="reference external" href="PhageHosts/code/crispr_blast2tax.py%20crispr_blast2tax.py">converted to
taxids</a>,
and <a class="reference external" href="PhageHosts/code/scoreTaxId.py%20scoreTaxId.py">scored</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/score_blast.py phage_with_host.fna.crispr.blastn best &gt; crispr.best.hits
python PhageHosts/code/crispr_blast2tax.py bas.best.hits &gt; crispr.taxid
python2.7 PhageHosts/code/scoreTaxId.py bas.taxid  &gt; bas.score
</pre></div>
</div>
</div>
<div class="section" id="exact-matches">
<h1>3. Exact matches<a class="headerlink" href="#exact-matches" title="Permalink to this headline">¶</a></h1>
<p>To identify the longest exact matches between the phage and bacteria we
took a two step approach. This is easier to code than the complete
solution, and although it takes slightly longer to run, the runtime is
not an important concern in this analysis.</p>
<p>We start by finding all 15mer hits between the bacteria and the phages,
and compiling those in order of their location. We then combine those
identical hits into longer stretches of similarity, because consecutive,
overlapping, hits must have come from two longer matching sequences.</p>
<p>To find all 15-mers in common between the phage and bacteria, sort and
join the matches, and list all the hits we use the following code. Note
that we need to find the matches in both the forward and reverse
direction, and then combine those matches.</p>
<div class="highlight-python"><div class="highlight"><pre>perl PhageHosts/code/find_exact_matches.pl ../phage_with_host.fna RefSeq/bacteria/complete_genomes.fna &gt; phage.15mers.bacteria.txt
perl PhageHosts/code/find_exact_matches.pl ../phage_with_host.rc.fna RefSeq/bacteria/complete_genomes.fna &gt; phage.15mers.bacteria.rc.txt
perl PhageHosts/code/sort_and_join_exact_matches.pl phage.15mers.bacteria.txt &gt; phage.kmers.bacteria.txt
perl PhageHosts/code/sort_and_join_exact_matches.pl phage.15mers.bacteria.rc.txt &gt;&gt; phage.kmers.bacteria.txt
perl PhageHosts/code/longest_exact_match.pl phage.kmers.bacteria.txt &gt; phage_best_hits.txt
</pre></div>
</div>
<p>to find the longest match per phage, tabulate the RefSeq IDs, and score
the matches we use:</p>
<div class="highlight-python"><div class="highlight"><pre>perl PhageHosts/code/longest_exact2tbl.pl  &gt; phage_best_hits_NCIDS.txt
python PhageHosts/code/NC2taxid.py phage_best_hits_NCIDS.txt  &gt; phage_best_hits_taxa.txt
python2.7 PhageHosts/code/scoreTaxId.py phage_best_hits_taxa.txt &gt; score.txt
</pre></div>
</div>
<p>To create the supplementary figure, we used</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 ../GitHubRepository/PhageHosts/code/exact_match_plot.py phage.kmers.bacteria.txt  &gt; phage_kmers.counts
</pre></div>
</div>
</div>
<div class="section" id="oligonucleotide-profiles">
<h1>4. Oligonucleotide profiles<a class="headerlink" href="#oligonucleotide-profiles" title="Permalink to this headline">¶</a></h1>
<div class="section" id="gc-content-of-coding-regions">
<h2>GC Content of Coding Regions<a class="headerlink" href="#gc-content-of-coding-regions" title="Permalink to this headline">¶</a></h2>
<p>Similar to codon usage below, we calculate the G+C/G+C+A+T content of
the coding regions. This gives us a simple one dimensional matrix that
we can use to map phages to hosts.</p>
<p>To count the GC content:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/cds_gc.py ../phage_with_host.cds.fna  &gt; phage.gc.tsv
</pre></div>
</div>
<p>To count the bacterial GC content and limit that to complete bacterial
sequences we used:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/cds_gc.py NCBI/RefSeq/bacteria/bacteria.$SGE_TASK_ID.orfs.fna.gz &gt; bacteria/bacteria.$SGE_TASK_ID.gc.tsv
python PhageHosts/code/complete_bacteria.py  &gt; complete_bacteria_gc.tsv
</pre></div>
</div>
<p>Now we have two files: <code class="docutils literal"><span class="pre">complete_bacteria_gc.tsv</span></code> and <code class="docutils literal"><span class="pre">phage.gc.tsv</span></code>
and we use pairwise <a class="reference external" href="http://en.wikipedia.org/wiki/Euclidean_distance">Euclidean
distance</a> to find
the closest organisms, convert them to taxa, and score the results.</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/cds_distance.py phage.gc.tsv complete_bacteria_gc.tsv  &gt; closest_genomes.ids
python PhageHosts/code/NC2taxid.py closest_genomes.ids &gt; closest_genomes.taxa
python PhageHosts/code/scoreTaxId.py closest_genomes.taxa &gt; score.txt
</pre></div>
</div>
</div>
<div class="section" id="codon-usage">
<h2>Codon usage<a class="headerlink" href="#codon-usage" title="Permalink to this headline">¶</a></h2>
<p>We count the codon profiles in both host and bacteria, and then
calculate the <a class="reference external" href="http://en.wikipedia.org/wiki/Euclidean_distance">Euclidean
distance</a> using
<a class="reference external" href="code/codon_distance.py">codon_distance.py</a> to identify the closest
host for each phage:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/codon_usage.py ../phage_with_host.cds.fna  &gt; phage_codon_counts.tsv
python codon_usage.py $DIR/bacteria.$SGE_TASK_ID.orfs.fna.gz &gt; bacteria/bacteria.$SGE_TASK_ID.codons.tsv
python PhageHosts/code/codon_distance.py phage_codon_counts.tsv bacteria_codon_counts.tsv  &gt; phage_bacteria_predictions.out
</pre></div>
</div>
<p>convert that output to taxonomy IDs and score the output:</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/NC2taxid.py phage_bacteria_predictions.out &gt; phage_bacteria_predictions.taxid
python2.7 PhageHosts/code/scoreTaxId.py phage_bacteria_predictions.taxid &gt; score.txt
</pre></div>
</div>
</div>
<div class="section" id="kmer-profiles">
<h2>Kmer profiles<a class="headerlink" href="#kmer-profiles" title="Permalink to this headline">¶</a></h2>
<p>We used <a class="reference external" href="http://www.genome.umd.edu/jellyfish.html">Jellyfish</a> to
count <em>k-</em>mers in the DNA sequences. The longest phage sequence is
358,663 bp, so we used a hash size of 400,000 to keep the whole thing
(or most of it) in memory.</p>
<p>Initially we started with this script to list all the 15-mers:</p>
<div class="highlight-python"><div class="highlight"><pre>for f in $(ls ../phage_with_host.fna.files_NC/);
    do n=$(echo $f | sed -e &#39;s/\.\.\/phage_with_host.fna.files_NC\///; s/fna$/tsv/&#39;);
    echo $f $n;
    jellyfish count -s 400000 -t 32 -C -m 15 -o 15mers.txt $f;
    jellyfish dump -ct 15mers.txt_0 &gt; 15mers/$n;
done
</pre></div>
</div>
<p>and then we modify it to count all k-mers between 3 and 20. That runs on
a node of the cluster in a few minutes.</p>
<p>We then combine those outputs to a single file using
<a class="reference external" href="code/combineKmerCounts.py">combineKmerCounts.py</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 PhageHosts/code/combineKmerCounts.py phage_kmer_counts/15mers 15mers.txt 0
</pre></div>
</div>
<p>That makes a single .txt file for each kmer size.</p>
<p>To count all bacterial kmers we need to extract the fasta sequences to
individual files and run jellyfish on all of those. The combination of
bash and perl does that:</p>
<div class="code bash highlight-python"><div class="highlight"><pre>BACTGZ=$ENV{GENOME}
BACT=$(echo $BACTGZ | sed -e &#39;s/.gz//&#39;)

if [ ! -e kmers ]; then mkdir kmers/; fi
if [ ! -e sequences ]; then mkdir sequences/; fi

mkdir sequences/
./split.pl $BACTGZ sequences/

mkdir kmers/
for FILE in $(ls sequences/); do
    for k in $(seq 3 10); do
            jellyfish count -s 400000 -t 32 -C -m $k -o kmers/$FILE.${k}kmer sequences/$FILE
            if [ -e kmers/$FILE.${k}kmer_1 ]; then
                    jellyfish merge -o kmers/$FILE.${k}kmer.output.jf kmers/$FILE.${k}kmer_*
            else
                    mv kmers/$FILE.${k}kmer_0 kmers/$FILE.${k}kmer.output.jf
            fi
            jellyfish dump -ct kmers/$FILE.${k}kmer.output.jf &gt; kmers/$FILE.${k}kmer.tsv
            rm -f kmers/$FILE.${k}kmer_*  kmers/$FILE.${k}kmer.output.jf
    done
done
</pre></div>
</div>
<p>This generates a single tsv file for every genome that we are going to
use in the analysis.</p>
<p>Finally, we just make sure we only include genomes that we are
interested in for our analysis.</p>
<div class="highlight-python"><div class="highlight"><pre>for i in $(ls all_phages); do echo $i; python trim.py all_phages/$i all_phages_trimmed/$i; done
for i in $(seq 3 9); do python trim.py bacterial_genomes/kmers/$i.kmers bacterial_kmers/$i.kmers; done
</pre></div>
</div>
</div>
</div>
<div class="section" id="cooccurrence-analysis">
<h1>5. Cooccurrence analysis<a class="headerlink" href="#cooccurrence-analysis" title="Permalink to this headline">¶</a></h1>
<p>We need to predict the presence of bacteria and phage in all of the
metagenomes that we have downloaded. We have 3,205 metagenomes that we
have downloaded from <a class="reference external" href="http://metagenomics.anl.gov/">MG-RAST</a>, however
we want to only use our complete set of phages and/or bacteria to do the
predictions, and not the random set of organisms that are present in
MG-RAST.</p>
<div class="section" id="predicting-bacterial-presence">
<h2>Predicting bacterial presence<a class="headerlink" href="#predicting-bacterial-presence" title="Permalink to this headline">¶</a></h2>
<p>We start with <a class="reference external" href="http://edwards.sdsu.edu/FOCUS/">FOCUS</a>, but we need to
make our unique databases first. Download the FOCUS code, and make the
databases like this. The final FOCUS command makes the bacterial
database.</p>
<div class="highlight-python"><div class="highlight"><pre>cp -r ~/bin/focus/bin focus_bacteria
cd focus_bacteria/db/
mv k6 k6_old
mv k7 k7_old
mv k8 k8_old
head -n 1 k6_old &gt; k6
head -n 1 k7_old &gt; k7
head -n 1 k8_old &gt; k8
cd ../..
grep -r \&gt; NCBI/RefSeq/bacteria/complete_genomes.fna.files.NC/  &gt; complete_genomes_files.txt
perl PhageHosts/code/focus_grep2list.pl &lt; complete_genomes_files.txt  &gt; complete_genomes_files.focus.txt
python2.7 focus_bacteria/focus.py -d complete_genomes_files.focus.txt
</pre></div>
</div>
<p>Now we can run all the bacterial genomes using focus. This script makes
a directory for each job, cd&#8217;s there, copies the data, runs focus, and
then cleans up.</p>
<p>This is focus.sh</p>
<div class="code bash highlight-python"><div class="highlight"><pre>#!/bin/bash

export LD_LIBRARY_PATH=:$HOME/lib
GZFILE=$(head -n  $SGE_TASK_ID mg-rast.txt | tail -n 1)
mkdir /scratch/$SGE_TASK_ID
cd /scratch/$SGE_TASK_ID
gunzip -c $GZFILE &gt; $SGE_TASK_ID.fasta
echo &quot;Input: &quot; $GZFILE  &gt; cooccurence/focus_NC/$SGE_TASK_ID.focus
python2.7 focus.py -m 0 -q $SGE_TASK_ID.fasta &gt;&gt; cooccurence/focus_NC/$SGE_TASK_ID.focus
cd $HOME/phage/host_analysis/cooccurence
rm -rf /scratch/$SGE_TASK_ID
</pre></div>
</div>
<p>Since there are 3025 files in mg-rast.txt, we use qsub to run this
command <code class="docutils literal"><span class="pre">qsub</span>&nbsp; <span class="pre">-cwd</span> <span class="pre">-S</span> <span class="pre">/bin/bash</span> <span class="pre">-t</span> <span class="pre">1-3025:1</span> <span class="pre">-o</span> <span class="pre">sge</span> <span class="pre">-e</span> <span class="pre">sge</span> <span class="pre">./focus.sh</span></code></p>
<p>This takes just a few hours to predict all the bacteria present in all
the metagenomes that have been produced, and once the computation is
complete, we combine them into a single file with the bacterial
predictions normalized to the lengths of the metagenomes:
<code class="docutils literal"><span class="pre">python</span> <span class="pre">~/bioinformatics/phage_host/focus_parse.py</span> <span class="pre">focus_NC2</span> <span class="pre">&gt;</span> <span class="pre">focus_bacteria_predictions.tsv</span></code></p>
</div>
<div class="section" id="predicting-phage-sequences">
<h2>Predicting phage sequences<a class="headerlink" href="#predicting-phage-sequences" title="Permalink to this headline">¶</a></h2>
<p>We have already run megablast for all phages against all metagenomes for
another project, and so we just use this as our input.</p>
<p>The code below makes two tables called <code class="docutils literal"><span class="pre">raw_phage_mg_counts.tsv</span></code> and
<code class="docutils literal"><span class="pre">normalized_phage_mg_counts.tsv</span></code> listing the presence of phages in
each of the metagenomes. The first is just the number of times that
phage is seen in each metagenome, and the second is the number divided
by the number of reads in the metagenome. It is the normalized number we
use (longer metagenomes have more hits, after all).</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 PhageHosts/code/count_phage_mg_hits.py 2&gt; err
</pre></div>
</div>
<p>Our computations produce two files, <code class="docutils literal"><span class="pre">normalized_phage_mg_counts.tsv</span></code>
and <code class="docutils literal"><span class="pre">focus_bacteria_predictions.tsv</span></code> that we need to compare. To do
so, we calculate Pearson&#8217;s correlation between genomes and phages,
convert those to taxonomy IDs and then score them.</p>
<div class="highlight-python"><div class="highlight"><pre>python PhageHosts/code/scoreDistances.py pearson_correlations.tsv max &gt; pearson.ncids
python PhageHosts/code/NC2taxid.py pearson.ncids &gt; pearson.tids
python2.7 PhageHosts/code/scoreTaxId.py pearson.tids &gt; pearson.score
</pre></div>
</div>
<p>We have also developed code to compare the scores using Euclidean
distance, Bray Curtis distance, City Block distance, Hamming distance,
and Jaccard distance, and we can run those sequentially, performing the
calculations, computing the taxonomy IDs and scoring the results.</p>
<div class="highlight-python"><div class="highlight"><pre>for i in euclidean braycurtis cityblock hamming jaccard;
    do python PhageHosts/code/scoreDistances.py ${i}_correlations.tsv min &gt; ${i}_correlations.ncids;
    python PhageHosts/code/NC2taxid.py ${i}_correlations.ncids &gt; ${i}_correlations.tids;
    python2.7 PhageHosts/code/scoreTaxId.py ${i}_correlations.tids &gt; ${i}_correlations.score;
done
</pre></div>
</div>
</div>
<div class="section" id="mutual-information">
<h2>Mutual Information<a class="headerlink" href="#mutual-information" title="Permalink to this headline">¶</a></h2>
<p>This is based on discussions with Peter Salamon, who said:</p>
<p>&#8220;You need a threshold of present in the metagenome to turn your current
variables into 0,1 variables that signal presence of the entity in the
metagenome. As a start, you could threshold on mean value for that
entity (phage or bacterium). Then you need the fraction of metagenomes
in which the phage and the bacterium co-occur, the fraction with one not
the other and the fraction with the other and not the one and the
fraction with neither. This gives the joint distribution of the four
cases. You also need the marginal distributions of fraction in which the
phage occurs and the fraction in which the bacterium occurs. Then simply
calculate the Kullback entropy of the joint distribution relative to the
product distribution.&#8221;</p>
<div class="highlight-python"><div class="highlight"><pre>python2.7 PhageHosts/code/mutual_information.py normalized_phage_mg_counts_scaled.tsv focus_bacteria_predictions.tsv 0 &gt; mi.mean.ncids
python PhageHosts/code/NC2taxid.py mi.mean.ncids &gt; mi.mean.tids
python2.7 PhageHosts/code/scoreTaxId.py mi.mean.tids &gt; mi.mean.score
</pre></div>
</div>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
</div>
</div>
<div class="section" id="you-can-also-find-what-you-are-looking-for-directly">
<h1>You can also find what you are looking for directly:<a class="headerlink" href="#you-can-also-find-what-you-are-looking-for-directly" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="#">Phage Host Analysis</a></li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2015, Robert Edwards and Bas Dutilh.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.3.1
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>